tscreen_posix.go

```diff
@@ -17,150 +17,65 @@
-// #ifdef B19200
-//	case B19200: return (19200);
-// #endif
-// #ifdef B38400
-//	case B38400: return (38400);
-// #endif
-// #ifdef B57600
-//	case B57600: return (57600);
-// #endif
-// #ifdef B76800
-//	case B76800: return (76800);
-// #endif
-// #ifdef B115200
-//	case B115200: return (115200);
-// #endif
-// #ifdef B153600
-//	case B153600: return (153600);
-// #endif
-// #ifdef B230400
-//	case B230400: return (230400);
-// #endif
-// #ifdef B307200
-//	case B307200: return (307200);
-// #endif
-// #ifdef B460800
-//	case B460800: return (460800);
-// #endif
-// #ifdef B921600
-//	case B921600: return (921600);
-// #endif
-//	}
-//	return (0);
-// }
-import "C"
-
-type termiosPrivate struct {
-	tios C.struct_termios
-}
+type termiosPrivate syscall.Termios
+
+var tiosp = termiosPrivate{}
 
-func (t *tScreen) termioInit() error {
-	var e error
-	var rv C.int
-	var newtios C.struct_termios
-	var fd C.int
+func (t *tScreen) termioInit() (err error) {
+	private := syscall.Termios{}
 
-	if t.in, e = os.OpenFile("/dev/tty", os.O_RDONLY, 0); e != nil {
-		goto failed
+	if t.in, err = os.OpenFile("/dev/tty", os.O_RDONLY, 0); err != nil {
+		return err
 	}
-	if t.out, e = os.OpenFile("/dev/tty", os.O_WRONLY, 0); e != nil {
-		goto failed
+	if t.out, err = os.OpenFile("/dev/tty", os.O_WRONLY, 0); err != nil {
+		return err
 	}
 
-	t.tiosp = &termiosPrivate{}
-
-	fd = C.int(t.out.Fd())
-	if rv, e = C.tcgetattr(fd, &t.tiosp.tios); rv != 0 {
-		goto failed
+	defer func() {
+		if err != nil {
+			if t.in != nil {
+				t.in.Close()
+			}
+			if t.out != nil {
+				t.out.Close()
+			}
+		}
+	}()
+
+	fd := t.out.Fd()
+
+	if err := termios.Tcgetattr(fd, &private); err != nil {
+		return fmt.Errorf("cannot get attributes: %s", err)
 	}
-	t.baud = int(C.getbaud(&t.tiosp.tios))
-	newtios = t.tiosp.tios
-	newtios.c_iflag &^= C.IGNBRK | C.BRKINT | C.PARMRK |
-		C.ISTRIP | C.INLCR | C.IGNCR |
-		C.ICRNL | C.IXON
-	newtios.c_oflag &^= C.OPOST
-	newtios.c_lflag &^= C.ECHO | C.ECHONL | C.ICANON |
-		C.ISIG | C.IEXTEN
-	newtios.c_cflag &^= C.CSIZE | C.PARENB
-	newtios.c_cflag |= C.CS8
+
+	t.baud = int(termios.Cfgetospeed(&private))
+
+	private.Iflag &^= unix.IGNBRK | unix.BRKINT | unix.PARMRK |
+		unix.ISTRIP | unix.INLCR | unix.IGNCR |
+		unix.ICRNL | unix.IXON
+	private.Oflag &^= unix.OPOST
+	private.Lflag &^= unix.ECHO | unix.ECHONL | unix.ICANON |
+		unix.ISIG | unix.IEXTEN
+	private.Cflag &^= unix.CSIZE | unix.PARENB
+	private.Cflag |= unix.CS8
 
 	// This is setup for blocking reads.  In the past we attempted to
 	// use non-blocking reads, but now a separate input loop and timer
 	// copes with the problems we had on some systems (BSD/Darwin)
 	// where close hung forever.
-	newtios.Cc[syscall.VMIN] = 1
-	newtios.Cc[syscall.VTIME] = 0
+	private.Cc[syscall.VMIN] = 1
+	private.Cc[syscall.VTIME] = 0
 
-	if rv, e = C.tcsetattr(fd, C.TCSANOW|C.TCSAFLUSH, &newtios); rv != 0 {
-		goto failed
+	if err = termios.Tcsetattr(fd, unix.TCSETS, &private); err != nil {
```

Comment:
What prompted the change to use TCSETS?

Reply:
So looking that the termios package, I think this was a mistake; the original values are the portable ones specified by the Open Group.  In fact, the termios package has specific knowledge to convert TCSANOW or TCSAFLUSH into non-portable values for some systems (like darwin).  I think this needs to be changed, or justified in more detail.

Reply:
x/sys/unix.TCSANOW Is not available on Solarish Platforms. TCSETS comes from another Package that uses Termios. AFAIk one of the examples of termios used it for the same purpose which is why I took that instead. If we should use TCSANOW we would need to define it manually.

Reply:
That seems like a bug in the termios package then -- TCSANOW should be defined everywhere.

Reply:
They are locally defined in the termios package, for example: https://github.com/pkg/term/blob/master/termios/termios_const_solaris.go

Reply:
In fact, you really should be using the values from that package, not from sys/unix.

Reply:
Yes you are right. Changed that. I also changed all others to be syscall instead of x/sys/unix. They are available there and i think it is better not to mix syscall.Termios struct with x/sys/unix Flags. If they have changed. Probably just a nitpick but lets better be save and only use one or the other and not mixed.